src/common/exif.cc

```diff
@@ -3568,6 +3570,54 @@ gboolean dt_exif_xmp_read(dt_image_t *img,
       goto end;
     }
     sqlite3_finalize(stmt);
+
+    /* Check if the XMP has the now default enabled highlights module.
+
+       If not we want to add this module but with the CLIP method instead of
+       the new default OPPOSED one. This is to keep compatibility with old-edits.
+     */
+
+    for(GList *iter = history_entries; iter; iter = g_list_next(iter))
+    {
+      history_entry_t *entry = (history_entry_t *)iter->data;
+
+      if(!strcmp(entry->operation, "highlights"))
+        has_highlights = TRUE;
+    }
+
+    // highlights module is not part of history, add a simple CLIP method
+    if(!has_highlights)
+    {
+      const char *default_clip =
+        "000000000000803f00000000000000000000803f"
+        "000000001e00000006000000cdcccc3e000000400000000000000000";
+      const char *no_blend = "gz11eJxjYGBgkGAAgRNODGiAEV0AJ2iwh+CRyscOAAdeGQQ=";
+
+      history_entry_t *entry = (history_entry_t *)calloc(1, sizeof(history_entry_t));
+      entry->operation = g_strdup("highlights");
+      entry->enabled = TRUE;
+      entry->modversion = 4;
+      entry->params = dt_exif_xmp_decode(default_clip,
+                                         strlen(default_clip),
+                                         &entry->params_len);
+      entry->multi_name = g_strdup("");
+      entry->multi_name_hand_edited = FALSE;
+      entry->multi_priority = 0;
+      entry->blendop_version = 13;
+      entry->blendop_params = dt_exif_xmp_decode(no_blend,
+                                                 strlen(no_blend),
+                                                 &entry->blendop_params_len);
+      // we insert the module in second position, just after
+      // rawprepare. This is to ensure that history_end do include
+      // this module. Just adding one to history_end won't work if the
+      // history_end is not pointing to the last history item.
+      entry->num = 1;
+      entry->iop_order = -1;
```

Comment:
I have no image with an xmp version < 2 for "real testing" but could you explain the '-1' for iop_order here?

Reply:
Because we don't care about the value here. Remember the iop_order is not part of the module data. So in the XMP we either are using one of the default iop order (tag `darktable:iop_order_version` - `DT_IOP_ORDER_LEGACY`,  `DT_IOP_ORDER_V30`, `DT_IOP_ORDER_V30_JPG` and highlights is listed in there) or we have a custom iop_order and so a list into the xmp (`darktable:iop_order_list`) and again highlights is listed in there. This is true for xmp_version >= 4. For older version the iop order list is reconstructed as it was not yet implemented and follows in this case the order of the module back then.

Hope this clarifies the situation. Do you think I may still have a missing code path?

Reply:
Thanks, it does clear my mind. And - no ideas about any missing path.